jetty-server/src/test/java/org/eclipse/jetty/server/AsyncContentProducerTest.java

```diff
@@ -0,0 +1,339 @@
+
+        Throwable error = readAndAssertContent(totalContentBytesCount, originalContentString, contentProducer, totalContentBytesCount + buffers.length + 2, 25, 4, barrier);
+        assertThat(error, nullValue());
+    }
+
+    @Test
+    public void testBlockingContentProducerGzipInterceptorWithError() throws Exception
+    {
+        ByteBuffer[] uncompressedBuffers = new ByteBuffer[3];
+        uncompressedBuffers[0] = ByteBuffer.wrap("1 hello 1".getBytes(StandardCharsets.ISO_8859_1));
+        uncompressedBuffers[1] = ByteBuffer.wrap("2 howdy 2".getBytes(StandardCharsets.ISO_8859_1));
+        uncompressedBuffers[2] = ByteBuffer.wrap("3 hey ya 3".getBytes(StandardCharsets.ISO_8859_1));
+        final int totalContentBytesCount = countRemaining(uncompressedBuffers);
+        final String originalContentString = asString(uncompressedBuffers);
+        final Throwable expectedError = new Throwable("HttpInput idle timeout");
+
+        ByteBuffer[] buffers = new ByteBuffer[3];
+        buffers[0] = gzipByteBuffer(uncompressedBuffers[0]);
+        buffers[1] = gzipByteBuffer(uncompressedBuffers[1]);
+        buffers[2] = gzipByteBuffer(uncompressedBuffers[2]);
+
+        CyclicBarrier barrier = new CyclicBarrier(2);
+
+        ContentProducer contentProducer = new AsyncContentProducer(new ArrayDelayedHttpChannel(buffers, new HttpInput.ErrorContent(expectedError), scheduledExecutorService, barrier));
+        contentProducer.setInterceptor(new GzipHttpInputInterceptor(inflaterPool, new ArrayByteBufferPool(1, 1, 2), 32));
+
+        Throwable error = readAndAssertContent(totalContentBytesCount, originalContentString, contentProducer, (buffers.length + 1) * 2, 0, 4, barrier);
+        assertThat(error, Is.is(expectedError));
+    }
+
+    private Throwable readAndAssertContent(int totalContentBytesCount, String originalContentString, ContentProducer contentProducer, int totalContentCount, int readyCount, int notReadyCount, CyclicBarrier barrier) throws InterruptedException, BrokenBarrierException, TimeoutException
+    {
+        int readBytes = 0;
+        String consumedString = "";
+        int nextContentCount = 0;
+        int isReadyFalseCount = 0;
+        int isReadyTrueCount = 0;
+        Throwable error = null;
+
+        while (true)
+        {
+            if (contentProducer.isReady())
+                isReadyTrueCount++;
+            else
+                isReadyFalseCount++;
+
+            HttpInput.Content content = contentProducer.nextContent();
+            nextContentCount++;
+            if (content == null)
+            {
+                barrier.await(5, TimeUnit.SECONDS);
+                content = contentProducer.nextContent();
+                nextContentCount++;
+            }
+            assertThat(content, notNullValue());
+
+            if (content.isSpecial())
+            {
+                if (content.isEof())
```

Comment:
Another case of "eof" implies "special"?

Reply:
Yes.